charms/kserve-controller/src/charm.py

```diff
@@ -0,0 +1,314 @@
+        self.framework.observe(
+            self.on.kube_rbac_proxy_pebble_ready, self._on_kube_rbac_proxy_ready
+        )
+
+        self._k8s_resource_handler = None
+        self._crd_resource_handler = None
+        self._lightkube_field_manager = "lightkube"
+        self._controller_container_name = "kserve-controller"
+        self.controller_container = self.unit.get_container(self._controller_container_name)
+
+        self._rbac_proxy_container_name = "kube-rbac-proxy"
+        self.rbac_proxy_container = self.unit.get_container(self._rbac_proxy_container_name)
+
+    @property
+    def _context(self):
+        """Returns a dictionary containing context to be used for rendering."""
+        self.gen_certs(self.model.name, self.app.name)
+        ca_context = b64encode(Path('/run/ca.crt').read_text().encode("ascii"))
+        return {
+            "app_name": self.app.name,
+            "namespace": self.model.name,
+            "cert": f"'{ca_context.decode('utf-8')}'",
+        }
+
+    @property
+    def k8s_resource_handler(self):
+        if not self._k8s_resource_handler:
+            self._k8s_resource_handler = KubernetesResourceHandler(
+                field_manager=self._lightkube_field_manager,
+                template_files=K8S_RESOURCE_FILES,
+                context=self._context,
+                logger=log,
+            )
+        return self._k8s_resource_handler
+
+    @property
+    def _controller_pebble_layer(self):
+        """Return the Pebble layer for the workload."""
+        return Layer(
+            {
+                "services": {
+                    self._controller_container_name: {
+                        "override": "replace",
+                        "summary": "KServe Controller",
+                        "command": "/manager --metrics-addr=:8080",
+                        "startup": "enabled",
+                        "environment": {
+                            "POD_NAMESPACE": self.model.name,
+                            "SECRET_NAME": "kserve-webhook-server-cert",
+                        },
+                    },
+                }
+            }
+        )
+
+    @property
+    def _rbac_proxy_pebble_layer(self):
+        """Return the Pebble layer for the workload."""
+        return Layer(
+            {
+                "services": {
+                    self._rbac_proxy_container_name: {
+                        "override": "replace",
+                        "summary": "Kube Rbac Proxy",
+                        "command": "/usr/local/bin/kube-rbac-proxy --secure-listen-address=0.0.0.0:8443 --upstream=http://127.0.0.1:8080 --logtostderr=true --v=10",
+                        "startup": "enabled",
+                    }
+                }
+            }
+        )
+
+    def _on_kserve_controller_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
```

Comment:
I think we should check container connection here before uploading anything:
`self.container.can_connect()` will return `True` if container available.

Reply:
I have added the [check](https://github.com/canonical/kserve-operators/blob/KF-803-kserve-refactor/charms/kserve-controller/src/charm.py#L132)